src/io/flutter/editor/ActiveEditorsOutlineService.java

```diff
@@ -0,0 +1,256 @@
+    }
+
+    // Remove obsolete outline listeners.
+    final List<String> obsoletePaths = new ArrayList<>();
+    synchronized (outlineListeners) {
+      for (final String path : outlineListeners.keySet()) {
+        if (!newPaths.contains(path)) {
+          obsoletePaths.add(path);
+        }
+      }
+      for (final String path : obsoletePaths) {
+        final FlutterOutlineListener listener = outlineListeners.remove(path);
+        if (listener != null) {
+          analysisServer.removeOutlineListener(path, listener);
+        }
+      }
+
+      // Register new outline listeners.
+      for (final String path : newPaths) {
+        if (outlineListeners.containsKey(path)) continue;
+        final FlutterOutlineListener listener = new OutlineListener();
+
+        outlineListeners.put(path, listener);
+        analysisServer.addOutlineListener(FileUtil.toSystemDependentName(path), listener);
+      }
+    }
+
+    synchronized (pathToOutline) {
+      for (final String path : obsoletePaths) {
+        // Clear the current outline as it may become out of date before the
+        // file is visible again.
+        pathToOutline.remove(path);
+      }
+    }
+    if (obsoletePaths.size() > 0 || newPaths.size() > 0) {
+      notifyEditorsUpdated();
+    }
+  }
+
+  private void notifyEditorsUpdated() {
+    for (Listener listener : listeners) {
+      listener.onEditorsChanged();
+    }
+  }
+
+  private void notifyOutlineUpdated(String path) {
+    for (Listener listener : listeners) {
+      listener.onOutlineChanged(path);
+    }
+  }
+
+  public void addListener(@NotNull Listener listener) {
+    listeners.add(listener);
+  }
+
+  public void removeListener(@NotNull Listener listener) {
+    listeners.remove(listener);
+  }
+
+  /**
+   * Gets the most up-to-date {@link FlutterOutline} for the file at {@param path}.
+   *
+   * <p>
+   * Note that {@link ActiveEditorsOutlineService#get(VirtualFile)} is a convenience method that gets the outline based on the
+   * {@link VirtualFile} directly.
+   */
+  @Nullable
+  public FlutterOutline get(String path) {
+    return pathToOutline.get(path);
+  }
+
+  /**
+   * Gets the most up-to-date {@link FlutterOutline} for the {@param file}.
+   */
+  @Nullable
+  public FlutterOutline get(VirtualFile file) {
+    return get(file.getCanonicalPath());
+  }
+
+  @Override
+  public void dispose() {
+    synchronized (outlineListeners) {
+      for (Map.Entry<String, FlutterOutlineListener> entry : outlineListeners.entrySet()) {
+        final String path = entry.getKey();
+        final FlutterOutlineListener listener = outlineListeners.remove(path);
+        if (listener != null) {
+          analysisServer.removeOutlineListener(path, listener);
+        }
+      }
+      outlineListeners.clear();
+    }
+    synchronized (pathToOutline) {
+      pathToOutline.clear();
+    }
+  }
+
+  /**
+   * Listener for changes to the active editors or open outlines.
+   */
+  public interface Listener {
```

Comment:
High level it would be cleaner if these events included the payload of editors and outlines rather than requiring clients to call methods back on the ActiveEditorsOutlineService object. That would reduce the risk of race conditions and perhaps simplify the public api for ActiveEditorsOutlineService.

Reply:
Done.  I was able to make the get method protected instead of public with this.